packages/vast-xml2js/src/helpers/xmlToJson.js

```diff
@@ -0,0 +1,66 @@
+const TYPE = {
+  CDATA: 'cdata',
+  DOCUMENT: 'document',
+  ELEMENT: 'element',
+  TEXT: 'text'
+};
+
+const nodeType = (node) => {
+  if (node.nodeType === 1) {
+    return TYPE.ELEMENT;
+  }
+
+  if (node.nodeType === 3 || node.nodeType === 4) {
+    return TYPE.TEXT;
+  }
+
+  if (node.nodeType === 9) {
+    return TYPE.DOCUMENT;
+  }
+
+  throw new Error('Unsupported element type');
+};
+
+const xmlToJson = (node) => {
+  const type = nodeType(node);
+
+  const obj = {
+    type
+  };
+
+  if (type === TYPE.ELEMENT) {
+    obj.name = node.nodeName.toLowerCase();
+
+    if (node.attributes.length > 0) {
+      obj.attributes = {};
+      for (const attribute of Array.from(node.attributes)) {
+        obj.attributes[attribute.nodeName] = attribute.nodeValue;
+      }
+    }
+  } else if (type === TYPE.TEXT) {
+    obj.text = node.nodeValue
+      .replace('<![CDATA[', '')
+      .replace(']]>', '')
+      .trim();
+  }
+
+  // do children
+  if (node.hasChildNodes()) {
```

Comment:
what would happen if try to do the content of the if, if there isn't any childNodes?

Reply:
not sure I understand. I there are no childNodes `hasChildNodes` will return false
https://developer.mozilla.org/en-US/docs/Web/API/Node/hasChildNodes

Reply:
if node.childNodes will be an empty like array then the code below don't need to be inside an if
if node.childNodes can be undefined and null then yes this needs to be inside this validation

Reply:
that is up to whoever implemented the browser logic. Not same to assume anything in my opinion.